argon2_out_of_process_support_sync.go

```diff
@@ -0,0 +1,247 @@
+// system-wide lock in this case because we don't hold a lock on the file that the lock
+// file path currently points to, so nothing prevents multiple processes from thinking
+// that they have taken the lock. Both of these cases can be tested for by doing the
+// following after acquiring an exclusive advisory lock on the open file descriptor for
+// the system lock file:
+// - Ensure that there is still a file at the system-wide lock file path.
+// - Ensure the inode that the system-wide lock file path currently points to matches
+//   the inode that we acquired an exclusive lock on.
+// If either of these checks fail, this process does not own the system-wide lock and
+// another attempt must be made to attempt to acquire it.
+//
+// For this to work reliably and without race conditions, the system-wide lock file must
+// only be unlinked by the current lock holder. Unlinking the current file would normally
+// happen when a process relinquishes the lock, but it's also ok to leave a stale lock
+// file around - in this case, another process will just re-use it rather than creating
+// a new one.
+
+var (
+	argon2SysLockStderr io.Writer = os.Stderr
+
+	errArgon2OutOfProcessHandlerSystemLockTimeout = errors.New("request timeout")
+)
+
+// acquireArgon2OutOfProcessHandlerSystemLock acquires the system-wide lock
+// for serializing Argon2 execution system-wide  via this package. If the
+// function returns with an error, then the lock was not acquired. If the
+// function returns wthout an error, the returned callback can be used to
+// release the lock (note that the lock will be relinquished automatically if
+// the process exits too).
+//
+// The specified timeout determines how long this function will wait before
+// aborting its attempt to acquire the lock. If set to 0, the function will
+// only perform a single attempt.
+func acquireArgon2OutOfProcessHandlerSystemLock(timeout time.Duration) (release func(), err error) {
+	var lockFile *os.File // The opened lock file
+
+	// Ensure that we close the open lockFile descriptor (if there is one)
+	// on error paths. Note that this does leave a lock file laying around,
+	// but this isn't a problem and unlinking it can only happen inside of
+	// an exclusive lock without breaking the locking contract anyway.
+	defer func() {
+		if err == nil || lockFile == nil {
+			return
+		}
+		if err := lockFile.Close(); err != nil {
+			fmt.Fprintf(argon2SysLockStderr, "Cannot close argon2 lock file descriptor on error: %v", err)
+		}
+	}()
+
+	timeoutTimer := time.NewTimer(timeout) // Begin the request timeout timer
+	triedOnce := false                     // Handle the case of timeout == 0
+
+	// Run a loop to try to acquire the lock.
+	for {
+		if triedOnce {
+			// If the loop has executed at least once, make sure that
+			// the timeout hasn't expired.
+			select {
+			case <-timeoutTimer.C:
+				// The timeout has expired.
+				return nil, errArgon2OutOfProcessHandlerSystemLockTimeout
```

Comment:
this case is not reached by tests

Reply:
It's a bit hit and miss whether the timeout on line 154 gets tested, or this one gets tested - it depends where the code lands before the timer expires. I'm not sure of a better way of doing it but I'll try and think of a way to make it a bit more reproducible.

Reply:
I've refactored this code to make it more reproducible where the code lands during tests.